examples/Reporting/ParallelReportDownload.cs

```diff
@@ -0,0 +1,242 @@
+        /// <exception cref="GoogleAdsException">Thrown if errors encountered in the execution of
+        ///     the request.</exception>
+        private async Task<SearchStreamStream> DownloadReportAsync(
+            GoogleAdsServiceClient googleAdsService, long customerId, string query, int reportId,
+            ConcurrentBag<ReportDownload> responses)
+        {
+            try
+            {
+                // Issue an asynchronous download request.
+                return await googleAdsService.SearchStreamAsync(
+                    customerId.ToString(), query,
+                    delegate(SearchGoogleAdsStreamResponse resp)
+                    {
+                        // Store the results.
+                        responses.Add(new ReportDownload(customerId, reportId, resp));
+                    }
+                );
+            }
+            catch (AggregateException ae)
+            {
+                Console.WriteLine($"Download failed for report {reportId}!");
+
+                GoogleAdsException gae = GoogleAdsException.FromTaskException(ae);
+
+                if (gae != null)
+                {
+                    Console.WriteLine($"Message: {gae.Message}");
+                    Console.WriteLine($"Failure: {gae.Failure}");
+                    Console.WriteLine($"Request ID: {gae.RequestId}");
+                    throw gae;
+                }
+
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Stores a result from a reporting API call. In this case we simply report a count of
+        /// the responses, but one could also write the response to a database/file, pass the
+        /// response on to another method for further processing, etc.
+        /// </summary>
+        private class ReportDownload
+        {
+            private long CustomerId { get; }
+            private int ReportId { get; }
+            private SearchGoogleAdsStreamResponse Response { get; }
+
+            /// <summary>
+            /// Initialize a saved report with a specific CID, an integer identifier, and the
+            /// returned response from a request.
+            /// </summary>
+            /// <param name="customerId">The customer ID for whom the report was requested.</param>
+            /// <param name="reportId">Integer to simplify identification of the request.</param>
+            /// <param name="response">The data received from a successful request.</param>
+            public ReportDownload(long customerId, int reportId,
+                SearchGoogleAdsStreamResponse response)
+            {
+                CustomerId = customerId;
+                ReportId = reportId;
+                Response = response;
+            }
+
+            public override string ToString()
+            {
+                return $"Results for report {ReportId} (CID {CustomerId}): " +
+                       $"{Response.Results.Count} rows returned.";
```

Comment:
Does this work? This requires that the entire report be downloaded, first, right?

Reply:
I believe it does for this specific scenario, and since it only stores a complete response.  A ReportDownload instance is only created once a download task has completed, and the printout only occurs once all tasks are complete.